// @sfpy/node-core/cjs/Error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SafepayUnknownError = exports.SafepayConflictError = exports.SafepayAuthenticationError = exports.SafepayAPIError = exports.SafepayInvalidRequestError = exports.SafepayError = void 0;
class SafepayError extends Error {
    constructor(raw = {}, type = null) {
        super(raw.message);
        this.type = type || this.constructor.name;
        this.status = raw.status;
        // @ts-ignore
        this.message = raw.message;
    }
}
exports.SafepayError = SafepayError;
/**
 * InvalidRequestError is raised when a request is initiated with invalid
 * parameters.
 */
class SafepayInvalidRequestError extends SafepayError {
    constructor(raw = {}) {
        super(raw, "SafepayInvalidRequestError");
    }
}
exports.SafepayInvalidRequestError = SafepayInvalidRequestError;
/**
 * APIError is a generic error that may be raised in cases where none of the
 * other named errors cover the problem. It could also be raised in the case
 * that a new error has been introduced in the API, but this version of the
 * Node.JS SDK doesn't know how to handle it.
 */
class SafepayAPIError extends SafepayError {
    constructor(raw = {}) {
        super(raw, "SafepayAPIError");
    }
}
exports.SafepayAPIError = SafepayAPIError;
/**
 * AuthenticationError is raised when invalid credentials are used to connect
 * to Safepay's servers.
 */
class SafepayAuthenticationError extends SafepayError {
    constructor(raw = {}) {
        super(raw, "SafepayAuthenticationError");
    }
}
exports.SafepayAuthenticationError = SafepayAuthenticationError;
/**
 * SafepayConflictError is raised when there is conflicting data on the server
 * e.g a user already exists
 */
class SafepayConflictError extends SafepayError {
    constructor(raw = {}) {
        super(raw, "SafepayConflictError");
    }
}
exports.SafepayConflictError = SafepayConflictError;
/**
 * Any other error from Safepay not specifically captured above
 */
class SafepayUnknownError extends SafepayError {
    constructor(raw = {}) {
        super(raw, "SafepayUnknownError");
    }
}
exports.SafepayUnknownError = SafepayUnknownError;
