// @sfpy/node-core/cjs/resources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = exports.Reporter = exports.Invoice = exports.Auth = exports.User = exports.Guests = exports.Payments = exports.Order = exports.Customers = void 0;
const ResourceNamespace_js_1 = require("./ResourceNamespace.js");
const Object_js_1 = require("./resources/Customers/Object.js");
const Object_js_2 = require("./resources/Guests/Object.js");
const Addresses_js_1 = require("./resources/Customers/Addresses.js");
const PaymentMethods_js_1 = require("./resources/Customers/PaymentMethods.js");
const Cancel_js_1 = require("./resources/Order/Cancel.js");
const Configure_js_1 = require("./resources/Order/Configure.js");
const Tracker_js_1 = require("./resources/Order/Tracker.js");
const Session_js_1 = require("./resources/Payments/Session.js");
const SignUp_js_1 = require("./resources/User/SignUp.js");
const Password_js_1 = require("./resources/User/Password.js");
const Object_js_3 = require("./resources/User/Object.js");
const Login_js_1 = require("./resources/Auth/Login.js");
const Logout_js_1 = require("./resources/Auth/Logout.js");
const Quicklinks_js_1 = require("./resources/Invoice/Quicklinks.js");
const Payments_js_1 = require("./resources/Reporter/Payments.js");
const Passport_js_1 = require("./resources/Client/Passport.js");
const Discount_js_1 = require("./resources/Order/Discount.js");
const Cards_js_1 = require("./resources/User/Cards.js");
const Vault_js_1 = require("./resources/Order/Vault.js");
exports.Customers = (0, ResourceNamespace_js_1.resourceNamespace)("customers", {
    Object: Object_js_1.Object,
    Addresses: Addresses_js_1.Addresses,
    PaymentMethods: PaymentMethods_js_1.PaymentMethods,
});
exports.Order = (0, ResourceNamespace_js_1.resourceNamespace)("order", {
    Tracker: Tracker_js_1.Tracker,
    Configure: Configure_js_1.Configure,
    Cancel: Cancel_js_1.Cancel,
    Discount: Discount_js_1.Discount,
    Vault: Vault_js_1.Vault,
});
exports.Payments = (0, ResourceNamespace_js_1.resourceNamespace)("payments", {
    Session: Session_js_1.Session,
});
exports.Guests = (0, ResourceNamespace_js_1.resourceNamespace)("guests", {
    Object: Object_js_2.Object,
});
exports.User = (0, ResourceNamespace_js_1.resourceNamespace)("user", {
    SignUp: SignUp_js_1.SignUp,
    Password: Password_js_1.Password,
    Object: Object_js_3.Object,
    Cards: Cards_js_1.Cards,
});
exports.Auth = (0, ResourceNamespace_js_1.resourceNamespace)("auth", {
    Login: Login_js_1.Login,
    Logout: Logout_js_1.Logout,
});
exports.Invoice = (0, ResourceNamespace_js_1.resourceNamespace)("invoice", {
    Quicklinks: Quicklinks_js_1.Quicklinks,
});
exports.Reporter = (0, ResourceNamespace_js_1.resourceNamespace)("reporter", {
    Payments: Payments_js_1.Payments,
});
exports.Client = (0, ResourceNamespace_js_1.resourceNamespace)("client", {
    Passport: Passport_js_1.Passport,
});
