// @sfpy/node-core/cjs/net/AxiosHttpClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AxiosHttpClientResponse = exports.AxiosHttpClient = void 0;
const axios_1 = require("axios");
const qs = require("qs");
const HttpClient_js_1 = require("./HttpClient.js");
class AxiosHttpClient extends HttpClient_js_1.HttpClient {
    /** @override. */
    getClientName() {
        return "node";
    }
    makeRequest(host, path, method, headers, requestData, requestParams, timeout) {
        return axios_1.default
            .create()
            .request({
            timeout: timeout,
            method: method,
            url: `${host}${path}`,
            headers: headers,
            data: requestData,
            params: requestParams,
            paramsSerializer: (params) => {
                return qs.stringify(params, {
                    arrayFormat: "indices",
                    encode: false,
                });
            },
        })
            .then((response) => {
            return new AxiosHttpClientResponse(response);
        })
            .catch((error) => {
            if (error.response) {
                const data = error.response.data;
                // @ts-ignore
                if (data.error) {
                    return new AxiosHttpClientResponse({
                        status: error.response.status,
                        headers: error.response.headers,
                        data: {
                            // @ts-ignore
                            message: data.error,
                        },
                    });
                    // @ts-ignore
                }
                else if (data.message) {
                    return new AxiosHttpClientResponse({
                        status: error.response.status,
                        headers: error.response.headers,
                        data: {
                            // @ts-ignore
                            message: data.message,
                        },
                    });
                }
                else {
                    return new AxiosHttpClientResponse({
                        status: error.response.status,
                        headers: error.response.headers,
                        data: {
                            // @ts-ignore
                            message: data.status.errors.join(", "),
                        },
                    });
                }
            }
            else if (error.request) {
                return new AxiosHttpClientResponse({
                    status: 500,
                    headers: {},
                    data: {
                        message: "The request was made but no response was received",
                    },
                });
            }
            else {
                return new AxiosHttpClientResponse({
                    status: 500,
                    headers: {},
                    data: {
                        message: "Something happened in setting up the request that triggered an Error",
                    },
                });
            }
        });
    }
}
exports.AxiosHttpClient = AxiosHttpClient;
class AxiosHttpClientResponse extends HttpClient_js_1.HttpClientResponse {
    constructor(res) {
        // @ts-ignore
        super(res.status, res.headers || {});
        this._res = res;
    }
    toJSON() {
        return new Promise((resolve) => resolve(this._res.data));
    }
}
exports.AxiosHttpClientResponse = AxiosHttpClientResponse;
